src/Gui/CommandFeat.cpp

```diff
@@ -86,29 +87,46 @@ void StdCmdRandomColor::activated(int iMsg)
 {
     Q_UNUSED(iMsg);
 
-    // get the complete selection
-    std::vector<SelectionSingleton::SelObj> sel = Selection().getCompleteSelection();
-
-    Command::openCommand(QT_TRANSLATE_NOOP("Command", "Set Random Color"));
-    for (const auto & it : sel) {
+    auto setRandomColor = [](ViewProvider* view) {
+        // NOLINTBEGIN
         auto fMax = (float)RAND_MAX;
         auto fRed = (float)rand()/fMax;
         auto fGrn = (float)rand()/fMax;
         auto fBlu = (float)rand()/fMax;
+        // NOLINTEND
         auto objColor = App::Color(fRed, fGrn, fBlu);
 
-        ViewProvider* view = Application::Instance->getDocument(it.pDoc)->getViewProvider(it.pObject);
         auto vpLink = dynamic_cast<ViewProviderLink*>(view);
-        if(vpLink) {
-            if(!vpLink->OverrideMaterial.getValue())
+        if (vpLink) {
+            if (!vpLink->OverrideMaterial.getValue()) {
                 vpLink->OverrideMaterial.setValue(true);
+            }
             vpLink->ShapeMaterial.setDiffuseColor(objColor);
-            continue;
         }
-        auto color = dynamic_cast<App::PropertyColor*>(view->getPropertyByName("ShapeColor"));
-        if (color) {
-            // get the view provider of the selected object and set the shape color
-            color->setValue(objColor);
+        else if (view) {
+            if (auto color = dynamic_cast<App::PropertyColor*>(view->getPropertyByName("ShapeColor"))) {
+                // get the view provider of the selected object and set the shape color
+                color->setValue(objColor);
+            }
+        }
+    };
+
+    // get the complete selection
+    std::vector<SelectionSingleton::SelObj> sel = Selection().getCompleteSelection();
+
+    Command::openCommand(QT_TRANSLATE_NOOP("Command", "Set Random Color"));
+    for (const auto & it : sel) {
+        ViewProvider* view = Application::Instance->getViewProvider(it.pObject);
+        setRandomColor(view);
+
+        Base::Type grpType = App::GroupExtension::getExtensionClassTypeId();
+        if (it.pObject->hasExtension(grpType)) {
+            auto *grp = static_cast<App::GroupExtension*>(it.pObject->getExtension(grpType));

```

Comment:
FIY there's a lint warning here:

> do not use static_cast to downcast from a base to a derived class; use dynamic_cast instead [cppcoreguidelines-pro-type-static-cast-downcast]

Reply:
This kind of linter warnings are annoying because in this context they are wrong. With the check that the object has an extension of type `GroupExtension` it's safe to downcast it with static_cast.

The recommended way by the cppcoreguidelines is to use dynamic_cast instead but then I do two type checks for the one and same thing. Furthermore, I have to also check the output of dynamic_cast as otherwise other code checkers will complain about that. All this makes the code to look unnecessarily complicated, then.

The only alternative I see ATM is to write (but not tested yet)
`if (auto grp = dynamic_cast<App::GroupExtension*>(it.pObject->getExtension(grpType, true, true)))`

But here we the problem of the so called _boolean blindness_ as for a code reviewer it's not obvious what the two true means. The reviewer has to check the API of `getExtension` to get an idea.

I think we need a template method of `getExtension` that allows to write elegant code. I will check this option.

Reply:
With 80d6d12ad1a I have added that template method. Now it can be used this way:
`if (auto grp = it.pObject->getExtension<App::GroupExtension>())`